### PR TITLE
Implement usage of `is_shardable` and `apply_sharding`

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61312 Sort imports of test_datapipe.py
* #61237 Adding backward compatibility for sharding support in old DataLoader
* **#61236 Implement usage of `is_shardable` and `apply_sharding`**
* #61235 Add DataPipes Graph Functions
* #61234 [DataLoader] Adding demux and mux DataPipe-s

Differential Revision: [D29588835](https://our.internmc.facebook.com/intern/diff/D29588835)

### CODE REVIEW

File: torch/utils/data/sharding.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+                results.add(item)
+        return results
+
+    all_pipes = traverse_graph(graph)
+
+    for pipe in all_pipes:
+        if hasattr(pipe, 'is_shardable'):
+            if pipe.is_shardable():
+                if hasattr(pipe, 'apply_sharding'):
+                    pipe.apply_sharding(num_of_instances, instance_id)
```
Comment (ejguan) [inline torch/utils/data/sharding.py:22]: I have a question here. Current test only takes one `shardable` datapipe. What if we have multiple DataPipe within pipeline, we should only make sharding happens once, right? If that is the case, considering the performance, sharding should happen as early as possible.

Comment (VitalyFedyunin) [reply torch/utils/data/sharding.py:22]: Good question, potentially we might have situation when A->B->C->D-> becomes 2 shards of B and 4 shards of C. But for sake of simple functionality at this particular moment, I will add exception if we shard more than once.

**Review by ejguan — APPROVED** [review]
LGTM

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29588835).

**Pull request closed without merging**